cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/InstanceOperationHttpHandler.java

```diff
@@ -0,0 +1,69 @@
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpRequest;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.util.Base64;
+import java.util.zip.ZipOutputStream;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+/**
+ * HTTP handler for instance level operations.
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class InstanceOperationHttpHandler extends AbstractHttpHandler {
+
+  private final CConfiguration cConf;
+  private final ApplicationLifecycleService lifecycleService;
+
+  @Inject
+  InstanceOperationHttpHandler(CConfiguration cConf, ApplicationLifecycleService lifecycleService) {
+    this.cConf = cConf;
+    this.lifecycleService = lifecycleService;
+  }
+
+  /**
+   * Exports all application details as a ZIP archive file.
+   */
+  @GET
+  @Path("/export/apps")
+  public void appsExport(HttpRequest request, HttpResponder responder) throws Exception {
+    File tempDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    DirUtils.mkdirs(tempDir);
```

Comment:
if there is a race between 2 RPC calls for export and the tempDir, this would fail right? May be we need to add a random suffix with directory name?

Reply:
No it won't. The `DirUtils.mkdirs` handle concurrent calls. We just need one tmp directory and have temp file (random name) created under it.

Reply:
Got it.